Zend/zend_vm_def.h

```diff
@@ -4486,18 +4485,24 @@ ZEND_VM_HANDLER(107, ZEND_CATCH, CONST, JMP_ADDR, LAST_CATCH|CACHE_SLOT)
 	}
 
 	exception = EG(exception);
-	ex = EX_VAR(opline->result.var);
-	if (UNEXPECTED(Z_ISREF_P(ex))) {
-		ex = Z_REFVAL_P(ex);
-	}
-	zval_ptr_dtor(ex);
-	ZVAL_OBJ(ex, EG(exception));
-	if (UNEXPECTED(EG(exception) != exception)) {
-		ZVAL_UNDEF(ex);
-		HANDLE_EXCEPTION();
+	if (RETURN_VALUE_USED(opline)) {
+		zval *ex = EX_VAR(opline->result.var);
+		if (UNEXPECTED(Z_ISREF_P(ex))) {
+			ex = Z_REFVAL_P(ex);
+		}
```

Comment:
Hm... not related to your patch, but I just realized that this can bypass typed references.

Reply:
This turned into something of a rabbit hole, but is now addressed with https://github.com/php/php-src/commit/314ab47e55845862be62683e331aa247025d073d and https://github.com/php/php-src/commit/4a08ca1294ea848fea2a0f87305d3b42226f16dd.